README.md
[![Build Status](https://travis-ci.org/ITV/domed-city.svg?branch=master)](https://travis-ci.org/ITV/domed-city)

# domed-city
Simple CLI application that wraps the Terraform API.

## Purpose

To consolidate, improve and enforce standards around ITV's use of Terraform across product teams.

This gem provides two main functions:

1. Wrap the main Terraform functions (e.g. `plan`, `apply`).
2. Ensure alignment to ITV's Common Platform VPC and environment policy.

## Disclaimer

This gem is very specific to how ITV utilise Terraform so it is unlikely to be useful to others except
to serve as an example of how we do things.

## Naming

From [Wikipedia](https://en.wikipedia.org/wiki/Domed_city):

> ...the dome is airtight and pressurized, creating a habitat that can be controlled for air temperature, composition and quality, typically due to an external atmosphere (or lack thereof) that is inimical to habitation for one or more reasons.

## Installation

Add to your Gemfile:

```
gem 'domed-city'
```

## Usage

For ease of use, type `bundle exec dome` (you may get some warnings if you do not use `bundle exec`) in the CLI:

```
$ bundle exec dome

  Dome wraps the Terraform API and performs useful stuff.

  Usage:
        dome [--sudo] --<command>
  where <command>s are:
  -p, --plan             Creates a Terraform plan
  -a, --apply            Applies a Terraform plan
  -i, --init             Run Terraform init
  -s, --state            Synchronises the Terraform state
  -o, --output           Print all Terraform output variables
  -t, --statecmd=<s+>    Wrapper arround terraform state
  -r, --refresh          Refresh the state
  -c, --console          Spawn terraform console
  -u, --sudo             Assume itv-root instead of the role specified in profile
  -j, --json             Pass the `-json` arg to plan and apply commands
  -v, --version          Print version and exit
  -h, --help             Show this message
  -e, --environment      Spawn sub-shell with the exported variables
```

Domed is designed to work with a certain directory structure. Your account,product,ecosystem and environment are assigned based on your current directory. The expected directory structure is terraform/$PRODUCT-$ECOSYSTEM/$ENVIRONMENT

It also requires certain things in itv.yaml.

1. Your project (or better product) is defined using the project key in your itv.yaml.
```
project: foo

```
2. Valid environments are defined using the hashmap of ecosystems to environments key in your itv.yaml.
eg
```
ecosystems:
  dev:
    - infradev
    - qa
    - stg
  prd:
    - infraprd
    - prd

```
3. Valid accounts are of the format <project>-dev and <project>-prd and are calculated automatically using your project variable. This is consistent with the .aws/config because dome will try and assume the role based on that account.

### Sudo mode

By default dome will assume the role defined in your AWS profile. If `--sudo` is specified, the value is overwritten by `itv-root` role in the same account.

Use this mode only when you need to manage resources requiring `itv-root`.

### Pin provider versions

Create a file called `.terraform-providers.yaml` in the root of the project:

```
aws: 2.6.0
external: 1.1.0
local: 1.2.0
template: 2.1.0
terraform: 1.0.0
vault: 1.6.0
```

## Development & Releases

In order to make changes, you can point the reference to domed-city in the Gemfile to your local working directory eg
```
gem 'domed-city', :path => '/home/foo/github-repos/domed-city'
```

* Every commit will trigger [travis CI](https://travis-ci.org/ITV/domed-city)
* Make sure you run rake spec & rake rubocop
* To release a new version make a PR with your code changes, update the CHANGELOG and lib/dome/version.rb with your version(semver). Then ask for someone to review your PR and merge it. Then manually create a github release.

## TODO

* Rename project references to product
* Check the usage of certificate section
* Remove dynamoDB state locking (Terraform does that now)
* When everyone moves to 1.1 uncomment and use the TF_VARs cidr_ecosystem_dev/prd


CHANGELOG.md
# 13.1.1
BUGFIX:
 - Fix release version

# 13.1.0
FEATURES:
- Adds functionality to get secrets from secrets manager for use as TF Vars before planning/applying with Dome
- Secrets can be configured in itv.yaml at the global, ecosystem and environment level such as:
```
dome:
 hiera_keys: {}
 certs: {}
 secretsmanager:
   global_secret: common_{ecosystem}_{environment}_secret_id
   dev:
     dev_common_secret: dev_common_secret_id
     dev:
       dev_dev_secret: dev_dev_secret_id
     qa:
       dev_qa_secret: dev_qa_secret_id
   prd:
     prd:
       prd_prd_secret: prd-prd_secret-id
```

- The secret id can use string replacement for {ecosystem} and {environment} so in the example above, planning in dev/qa would result in global_secret returning the secret string for the common_dev_qa_secret_id secret in the relevant account.
- Secrets with duplicate names are overwritten by the most precise scope (env > eco > global)
- Secret names are output as environment variables with the TF_VAR_ prefix, as per env vars set via Hiera.
- Secrets can only be access from the authenticated AWS account.

# 13.0.4
BUGFIX:
- upload kpi changes in CI
# 13.0.3
BUGFIX:
- check ecosystem before aws-vault kpi commands
# 13.0.2
BUGFIX:
- cleanup kpi csv file after upload
# 13.0.1
BUGFIX:
- correct bucket/prefix referenced to get version ID
- ensure breaking changes in 13.0.0 are adhered to
# 13.0.0
BREAKING:
- Add prd infra changes to core kpi bucket
- Requires tf-aws-eksci v32.0.0
- Requires cp-jenkins-libs v12.0.0

# 12.4.0
FEATURE:
- Add support for darwin arm64 providers

# 12.3.0
FEATURE:
- add optional `--json` arg to use `-json` in plan/apply

# 12.2.1
BUGFIX:
- add params at ecosystem level during import

# 12.2.0
FEATURES:
  - support installing 3rd-party providers.
  - support providing optional providers via a `.terraform-providers-local.yaml` file.

# 12.1.0
FEATURES:
  - Update the ecoroles config to use directory name `roles` at the ecosystem level
# 12.0.1
PATCH:
  - Check time left on sso CLI sessions

# 12.0.0
BREAKING:
  - Read params/env.tfvars at the ecosystem level.
  - This version should only be used when updating to the `iron` release

# 11.1.0
FEATURE:
  - Remove old tf12-13 upgrade provider cache
BUGFIX:
  - Use new providercache dir for tf14 onwards

# 11.0.0
BREAKING:
  - Update providers for namespace functionality as part of TF14 update
  - Remove TF lock file when running TF plan
  - This version should only be used when updating to TF14 (chromium release)

# 10.4.1
BUGFIX:
  - Set `AWS_REGION` envvar rather than `AWS_DEFAULT_REGION`
    - this is to fix a [provider bug](https://github.com/hashicorp/terraform-provider-aws/issues/5981) that prevents you from creating S3 buckets outside of `eu-west-1`

# 10.4.0
FEATURE:
  - Show remaining seconds left on session
  - Fix typo

# 10.3.0
FEATURE:
  - Export Main Environment when running actions at Ecosystem level to support the environment to ecosystem migration

# 10.2.0
FEATURE:
  - Check AWS session time before running. If the remaining session time is <20 minutes then clear the session and get a new one. If less than 40 minutes, check if the user wants to clear the session before continuing. This is for the case of long-running applys such as EKS upgrades.

# 10.1.1
FEATURE:
  - Load secrets into env vars for terraform import

# 10.1.0
FEATURE:
  - Add terraform import option

# 10.0.0
BREAKING:
  - Update the providers download to support the new local provider path. Also provides a second compatibility provider dir for when migrating from tf1
2 -> tf13. This version should only be used when updating to TF13 (scandium release)

# 9.2.1
BUGFIX:
- Only warn on missing hiera lookups

# 9.2.0
FEATURE:
- Add `ecoroles` level at the ecosystem

# 9.1.1
BUGFIX:
- Handle outputs of `aws-vault` on Linux clients

# 9.1.0
FEATURE:
- Check terraform version against .terraform-version before running

# 9.0.0
BREAKING:

- Upgrade aws-sdk to v3
- This requires the following updates in your infra's Gemfile:
  - Update: gem 'aws-sdk-core', '~> 3'
  - Update lugus gem to at least 'v9.0.0'
  - Remove: gem 'aws_assume_role', '1.2.1'

BUGFIX:
- Fix zip file open for use in containers

# 8.0.0

BREAKING CHANGES:
- Remove `aws_assume_role` gem
- Use `aws-vault` for assuming role
  - Install it separately: https://github.com/99designs/aws-vault
  - New AWS config format: Uses `<product>-<ecosysyem>-[dev|pe|root]` profile names
  - If you want to use `itv-dev` role by default put `export ITV_DEV=true` into your shell's rc file
  - To skip invoking `aws-vault` run `export FREEZE_AWS_ENVVAR=true`
- To use Yubikey for MFA, set the following environment variable (with your email!):
  - `export YUBIKEY_MFA='Amazon Web Services:first.last@itv.com@itv-root'`
  - you must have ykman installed for this to work: https://developers.yubico.com/yubikey-manager/

# 7.1.0
FEATURES:
- Pin to Ruby 2.7.1
- Update vulnerable Rake

# 7.0.3
FEATURES:
- Allow dome to use `params/env.tf` from the environment level when running on new run-level `services`

# 7.0.2
FEATURES:
- Add `service` level where a business service can be defined within `<product>-infra/terraform/<product>-<ecosystem>/<environment>/<services>/<serviceX>`. This ensures a one services AWS S3 bucket with multiple uniquely named terraform state files.
- Add explicit validation for `services` and `roles` levels.

`NOTE: Do not use v7.0.0 or v7.0.1 as v7.0.2 contains the necessary incremental fixes.`

# 7.0.1
FEATURES:
- Provide better naming convention for service-level state file i.e. `<project>-<ecosystem>-<environment>-<services>.tfstate`.

# 7.0.0
FEATURES:
- Add `service` level where a business service can be defined within `<product>-infra/terraform/<product>-<ecosystem>/<environment>/<service>`. This ensures a service-specific AWS S3 bucket and terraform state file.

# 6.18.2

FEATURES:
- Provide clearer error if required profile is missing from aws config

# 6.18.1

BUGFIX:
- Add `rubyzip` dependency


lib/dome/environment.rb
# frozen_string_literal: true

# This class represents the current directory

require 'open3'
require 'io/console'

module Dome
  class Environment
    attr_reader :environment, :account, :settings, :services

    include Dome::Level

    def initialize(directories = Dir.pwd.split('/'))
      ENV['AWS_REGION'] = 'eu-west-1'

      puts <<-'MSG'
             _
          __| | ___  _ __ ___   ___
        /  _` |/ _ \| '_ ` _ \ / _ \
        | (_| | (_) | | | | | |  __/
         \__,_|\___/|_| |_| |_|\___|

        Wrapping terraform since 2015
      MSG

      puts ''
      puts "[*] Operating at #{level.colorize(:red)} level"
      puts ''

      @sudo = false

      @settings               = Dome::Settings.new
      @product                = @settings.parse['product']

      case level
      when 'environment'
        @environment            = directories[-1]
        @account                = directories[-2]
        @services               = nil

      when 'ecosystem'
        @environment            = directories[-1].split('-')[-1]
        @account                = directories[-1]
        @services               = nil

      when 'ecoroles'
        @environment            = directories[-2].split('-')[-1]
        @account                = directories[-2]
        @services               = nil

      when 'product'
        @environment            = nil
        @account                = "#{@product}-prd"
        @services               = nil

      when 'roles'
        @environment            = directories[-2]
        @account                = directories[-3]
        @services               = nil

      when 'services'
        @environment            = directories[-3]
        @account                = directories[-4]
        @services               = directories[-1]

      when 'ecoservices'
        @environment            = directories[-3].split('-')[-1]
        @account                = directories[-3]
        @services               = directories[-1]

      when 'ecoservices-all'
        @environment            = directories[-2].split('-')[-1]
        @account                = directories[-2]
        @services               = directories[-1]

      when 'services-all'
        @environment            = directories[-2]
        @account                = directories[-3]
        @services               = directories[-1]

      when /^secrets-/
        @environment            = directories[-3]
        @account                = directories[-4]
        @services               = nil

      else
        puts "Invalid level: #{level}".colorize(:red)
      end

      @ecosystem              = @account.split('-')[-1]
      @account_id             = @settings.parse['aws'][@ecosystem.to_s]['account_id'].to_s

      ENV['TF_VAR_product']   = @product
      ENV['TF_VAR_envname']   = @environment
      ENV['TF_VAR_env']       = @environment
      ENV['TF_VAR_ecosystem'] = @ecosystem
      ENV['TF_VAR_aws_account_id'] = @account_id

      cidr_ecosystem = []
      cidr_ecosystem_dev = []
      cidr_ecosystem_prd = []

      ecosystem_environments = @settings.parse['aws'][@ecosystem.to_s]['environments'].keys
      ecosystem_environments.each do |k|
        cidr_ecosystem << @settings.parse['aws'][@ecosystem.to_s]['environments'][k.to_s]['aws_vpc_cidr']
      end

      dev_ecosystem_environments = @settings.parse['aws']['dev']['environments'].keys
      dev_ecosystem_environments.each do |k|
        cidr_ecosystem_dev << @settings.parse['aws']['dev']['environments'][k.to_s]['aws_vpc_cidr']
      end

      prd_ecosystem_environments = @settings.parse['aws']['prd']['environments'].keys
      prd_ecosystem_environments.each do |k|
        cidr_ecosystem_prd << @settings.parse['aws']['prd']['environments'][k.to_s]['aws_vpc_cidr']
      end

      ENV['TF_VAR_cidr_ecosystem'] = cidr_ecosystem.join(',').to_s

      #
      # TODO: Will uncomment when all the products migrate to 1.1
      #

      # ENV['TF_VAR_cidr_ecosystem_dev'] = cidr_ecosystem_dev.join(',').to_s
      # ENV['TF_VAR_cidr_ecosystem_prd'] = cidr_ecosystem_prd.join(',').to_s

      ENV['TF_VAR_dev_ecosystem_environments'] = dev_ecosystem_environments.join(',').to_s
      ENV['TF_VAR_prd_ecosystem_environments'] = prd_ecosystem_environments.join(',').to_s

      puts '--- Initial TF_VAR variables to drive terraform ---'
      puts "[*] Setting aws_account_id to #{ENV['TF_VAR_aws_account_id'].colorize(:green)}"
      puts "[*] Setting product to #{ENV['TF_VAR_product'].colorize(:green)}"
      puts "[*] Setting ecosystem to #{ENV['TF_VAR_ecosystem'].colorize(:green)}"
      puts "[*] Setting env to #{ENV['TF_VAR_env'].colorize(:green)}" unless ENV['TF_VAR_env'].nil?
      puts "[*] Setting cidr_ecosystem to #{ENV['TF_VAR_cidr_ecosystem'].colorize(:green)}"
      puts ''

      puts '--- The following TF_VAR are helpers that modules can use ---'
      puts "[*] Setting dev_ecosystem_environments to #{ENV['TF_VAR_dev_ecosystem_environments'].colorize(:green)}"
      puts "[*] Setting prd_ecosystem_environments to #{ENV['TF_VAR_prd_ecosystem_environments'].colorize(:green)}"

      #
      # TODO: Will uncomment when all the products migrate to 1.1
      #

      # puts "[*] Setting cidr_ecosystem_dev to #{ENV['TF_VAR_cidr_ecosystem_dev'].colorize(:green)}"
      # puts "[*] Setting cidr_ecosystem_prd to #{ENV['TF_VAR_cidr_ecosystem_prd'].colorize(:green)}"

      puts ''
    end

    def project
      @settings.parse['project']
    end

    def ecosystem
      directories = Dir.pwd.split('/')
      case level
      when 'ecosystem'
        directories[-1].split('-')[-1]
      when 'ecoroles'
        directories[-2].split('-')[-1]
      when 'environment'
        directories[-2].split('-')[-1]
      when 'product'
        # FIXME: This is 'prd' if accessed as @ecosystem
        'product'
      when 'roles'
        directories[-3].split('-')[-1]
      when /^secrets-|services/
        directories[-4].split('-')[-1]
      else
        puts "Invalid level: #{level}".colorize(:red)
      end
    end

    def accounts
      %W[#{project}-dev #{project}-prd]
    end

    def environments
      ecosystems = @settings.parse['ecosystems']
      raise '[!] ecosystems must be a hashmap of ecosystems to environments' unless ecosystems.is_a?(Hash)

      ecosystems.values.flatten
    end

    def check_session_time(profile)
      # Check remaining session time
      list_last_stdout, wait_threads = Open3.pipeline_r("aws-vault list")
      list_output = list_last_stdout.read
      status = wait_threads.first.value
      unless status.success?
        puts 'Could not determine AWS session time remaining. Continuing anyway'.colorize(:red)
        return true
      end

      session_left = ''
      list_output.each_line do |line|
        if line.split[0] == profile
          session_left = line.split[2]
        end
      end

      if session_left.start_with?('sts.AssumeRole:') || session_left.start_with?('sso.GetRoleCredentials:')
        timeleft = session_left.split(':')[1]
        timeleft = "0m#{timeleft}" if ! timeleft.include? 'm'
        timeleft = "0h#{timeleft}" if ! timeleft.include? 'h'
        timeleft = '0h0m0s' if timeleft.start_with?('-')
        secs_left = Time.parse(timeleft) - Time.parse('0h0m0s')
        if secs_left < 1200 # 20 mins
          puts "Less than 20 minutes AWS session time remaining (#{Time.at(secs_left).utc.strftime("%Mm %Ss")}). Clearing session".colorize(:red)
          return false
        end
        if secs_left < 2400 # 40 mins
          puts "Less than 40 minutes AWS session time remaining (#{Time.at(secs_left).utc.strftime("%Mm %Ss")}). Start a new session(y|N)?".colorize(:yellow)
          prompt = STDIN.getch
          return false if prompt.downcase == 'y'
        end
      end

      return true
    end

    def clear_session(profile)
      stdout, wait_threads = Open3.pipeline_r("aws-vault clear #{profile}")
      status = wait_threads.first.value
      unless status.success?
        puts output.colorize(:red)
        raise 'Unable to clear aws-vault session'
      end
    end

    def aws_credentials

      if ENV['FREEZE_AWS_ENVVAR']
        puts '$FREEZE_AWS_ENVVAR is set. Leaving AWS environment variables unchanged.'
        return
      end

      if ENV['YUBIKEY_MFA']
        cmd_yubikey = "ykman oath code --single '#{ENV['YUBIKEY_MFA']}'"
        last_stdout, wait_threads = Open3.pipeline_r(cmd_yubikey)
        mfa = last_stdout.read
        status = wait_threads.first.value

        unless status.success?
          puts mfa.colorize(:red)
          raise 'Unable to assume role'
        end
      end

      profile_suffix = 'pe'
      profile_suffix = 'dev' if ENV['ITV_DEV']
      profile_suffix = 'root' if @sudo
      profile = "#{account}-#{profile_suffix}"

      cmd_vault = "aws-vault exec #{profile} -- env"
      cmd_vault = "aws-vault exec #{profile} -t #{mfa.strip} -- env" if ENV['YUBIKEY_MFA']

      valid_session = false
      while !valid_session do
        last_stdout, wait_threads = Open3.pipeline_r(cmd_vault)
        output = last_stdout.read
        status = wait_threads.first.value

        unless status.success?
          puts output.colorize(:red)
          raise 'Unable to assume role'
        end
        valid_session = check_session_time(profile)
        clear_session(profile) unless valid_session
      end

      env = output.split("\n").grep(/^AWS/).map { |var| Hash[*var.split('=', 2)] }.reduce({}, &:merge)

      puts '[*] Exporting temporary credentials to environment variables '\
      "#{'AWS_ACCESS_KEY_ID'.colorize(:green)}, #{'AWS_SECRET_ACCESS_KEY'.colorize(:green)}"\
      " and #{'AWS_SESSION_TOKEN'.colorize(:green)}."
      ENV['AWS_ACCESS_KEY_ID'] = env['AWS_ACCESS_KEY_ID']
      ENV['AWS_SECRET_ACCESS_KEY'] = env['AWS_SECRET_ACCESS_KEY']
      ENV['AWS_SESSION_TOKEN'] = env['AWS_SESSION_TOKEN']
      puts ''
    end

    def valid_account?(account_name)
      accounts.include? account_name
    end

    def valid_environment?(environment_name)
      environments.include? environment_name
    end

    def invalid_account_message
      generic_error_message
      raise "\n[!] '#{@account}' is not a valid account.\n".colorize(:red)
    end

    def invalid_environment_message
      generic_error_message
      raise "\n[!] '#{@environment}' is not a valid environment.\n".colorize(:red)
    end

    def sudo
      @sudo = true
    end

    private

    def generic_error_message
      puts ''
      puts '--- Debug --- '
      puts "The environments you have defined are: #{environments}."
      puts "The accounts we calculated from your project itv.yaml key are: #{accounts}."
      puts ''
      puts '--- Troubleshoot ---'
      puts 'To fix your issue, try the following:'
      puts '1. Set your .aws/config to one of the valid accounts above.'
      puts '2. Ensure you are running this from the correct directory.'
      puts '3. Update your itv.yaml with the required environments or project.'
      puts '4. Check the README in case something is missing from your setup or ask in Slack'
      puts ''
    end
  end
end
